hedera-node/hedera-app/src/main/java/com/hedera/node/app/records/impl/producers/StreamFileProducerConcurrent.java

```diff
@@ -182,64 +197,79 @@ public void switchBlocks(
+            // serialized items to record file in parallel update running hash in a background thread
+            lastRecordHashingResultNMinus3 = lastRecordHashingResultNMinus2;
+            lastRecordHashingResultNMinus2 = lastRecordHashingResultNMinus1;
+            lastRecordHashingResultNMinus1 = lastRecordHashingResult;
+            lastRecordHashingResult = lastRecordHashingResult
+                    .thenCombine(futureSerializedRecords, TwoResults::new)
+                    .thenApplyAsync(
+                            twoResults -> format.computeNewRunningHash(twoResults.a(), twoResults.b()),
+                            executorService);
+            // write serialized items to record file in a background thread
+            currentRecordFileWriter = currentRecordFileWriter
+                    .thenCombine(futureSerializedRecords, TwoResults::new)
+                    .thenApplyAsync(
+                            twoResults -> {
+                                final var writer = twoResults.a();
+                                final var serializedItems = twoResults.b();
+                                serializedItems.forEach(item -> {
+                                    try {
+                                        writer.writeItem(item);
+                                    } catch (final Exception e) {
+                                        logger.error("Error writing record item to file", e);
+                                    }
+                                });
+                                return writer;
+                            },
+                            executorService);
+        } finally {
+            lock.unlock(); // Always unlock.
+        }
     }
 
     /**
      * Closes this StreamFileProducerBase wait for any background thread, close all files etc. This method is
-     * synchronous, so waits for all background threads to finish and files to be closed.
+     * synchronous, so waits for all background threads to finish and files to be closed. This method is thread-safe and
+     * can be called outside the context of the handle thread.
      */
     @Override
     public void close() {
-        if (currentRecordFileWriter != null) {
-            CompletableFuture.allOf(currentRecordFileWriter, lastRecordHashingResult)
-                    .thenAccept(aVoid -> {
-                        final var writer = currentRecordFileWriter.join();
-                        final var lastRunningHash = lastRecordHashingResult.join();
-                        closeWriter(writer, lastRunningHash);
-                    })
-                    .join();
+        lock.lock(); // Block until the lock is acquired
+
+        try {
+            if (currentRecordFileWriter != null) {
+                CompletableFuture.allOf(currentRecordFileWriter, lastRecordHashingResult)
+                        .thenAccept(aVoid -> {
+                            final var writer = currentRecordFileWriter.join();
```

Comment:
These `join` calls should probably be `get` calls to avoid unnecessary thread switching.

Reply:
Yes, there are future-related concepts in this class that need optimizing but that's outside the scope of what this PR is trying to fix. Good point for a follow up PR.